src/gdb.ts

```diff
@@ -68,11 +64,47 @@ class GDBDebugSession extends MI2DebugSession {
 			this.isSSH = true;
 			this.trimCWD = args.cwd.replace(/\\/g, "/");
 			this.switchCWD = args.ssh.cwd;
-			this.miDebugger.ssh(args.ssh, args.ssh.cwd, args.target, args.arguments, args.terminal, false).then(() => {
-				if (args.autorun)
-					args.autorun.forEach(command => {
-						this.miDebugger.sendUserInput(command);
-					});
+		}
+		if (args.autorun === undefined)
+			args.autorun = [];
+
+		if (!hasAutorunBeforeArgs)
+			args.autorunBefore = ["gdb-set target-async on",
+				"environment-directory \"$cwd\""];
+
+		if (attach) {
+			this.attached = !args.remote;
+			if (!hasAutorunBeforeArgs)
+				args.autorunBefore.push("target-select remote $target");
+
+		} else {
+			this.attached = false;
+			this.needContinue = true;
+			if (!hasAutorunBeforeArgs)
+				args.autorunBefore.push("target-select $target");
+		}
+		args.autorun = args.autorun.map(
+			s => {return escape(s);
+		});
+
+		args.autorunBefore = args.autorunBefore.map(
```

Comment:
don't use escape for this, it will put a `\` before all other `\` and `"` characters, this will manipulate user input and it will break all commands using `"` or `\`. I'm not quite sure what you mean with the workspaceroot.